eng-wasm/derive/src/pub_interface/parse_signatures.rs

```diff
@@ -0,0 +1,600 @@
+#[derive(Fail, Debug)]
+enum ParseError {
+    #[fail(display = "pub_interface item must be either a trait or an inherent struct impl")]
+    BadInputItem,
+
+    #[fail(display = "pub_interface item can not be a trait impl")]
+    TraitImpl,
+
+    #[fail(display = "pub_interface item can not have methods that receive `self`")]
+    MethodWithReceiver,
+
+    #[fail(
+        display = "methods in traits annotated with pub_interface should not define default implementations"
+    )]
+    TraitMethodWithImplementation,
+
+    #[fail(
+        display = "methods in traits annotated with pub_interface should not be annotated with additional attributes"
+    )]
+    TraitMethodWithAttributes,
+
+    #[fail(
+        display = "methods in impls annotated with pub_interface should not be annotated with no_mangle"
+    )]
+    ImplMethodWithNoMangleAttribute,
+
+    #[fail(
+        display = "methods in impls annotated with pub_interface should be either `fn` or `pub fn`"
+    )]
+    ImplMethodWithBadVisibility,
+}
+
+enum PubInterfaceInput {
+    ItemTrait(syn::ItemTrait),
+    ItemImpl(syn::ItemImpl),
+}
+
+/// The signatures collected while parsing the macro input
+#[allow(dead_code)]
+#[derive(Clone)]
+pub(crate) struct PubInterfaceSignatures {
+    /// The path to the type implementing the methods.
+    pub(crate) implementor: syn::Type,
+
+    /// The list of exported signatures
+    pub(crate) signatures: Vec<syn::Signature>,
+}
+
+impl syn::parse::Parse for PubInterfaceSignatures {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let mut forked_input;
+
+        // Speculatively try to parse the input in several ways.
+
+        // First, try to parse the input as a trait,
+        // which will  be backwards compatible with the original API...
+        let pub_interface_input = {
+            forked_input = input.fork();
+            forked_input
+                .parse::<syn::ItemTrait>()
+                .map(PubInterfaceInput::ItemTrait)
+        }
+        // If that fails, try to parse the input as an impl of a struct/enum...
+        .or_else(|_err| {
+            forked_input = input.fork();
+            forked_input
+                .parse::<syn::ItemImpl>()
+                .map(PubInterfaceInput::ItemImpl)
+        })
+        // If one of the attempts worked,
+        // signal to the parse mechanism that parsing finished successfully.
+        .map(|pub_interface_input| {
+            input.advance_to(&forked_input);
+            pub_interface_input
+        })
+        // If none of the options worked, we tell the user that he gave us bad input.
+        .map_err(|_err| input.error(ParseError::BadInputItem))?;
+
+        let mut implementor: syn::Type;
+        let signatures = match pub_interface_input {
+            PubInterfaceInput::ItemTrait(item_trait) => {
+                let default_implementor_name = super::DEFAULT_IMPLEMENTOR_NAME.into_ident();
```

Comment:
Please pay attention: the fixed name `DEFAULT_IMPLEMENTOR_NAME` of the structure is a *defect* that you luckily revealed. Any name should be allowed. I will appreciate if you can add this fix as well.

Reply:
fixed :)